python/dolfinx/fem/assemble.py

```diff
@@ -213,7 +213,7 @@ def _assemble_vector_array(b: np.ndarray, L: Form, constants=None, coeffs=None):
 
 
 @functools.singledispatch
-def assemble_matrix(a: typing.Any, bcs: typing.Optional[typing.List[DirichletBC]] = None,
+def assemble_matrix(a: typing.Any, bcs: typing.Optional[list[DirichletBC]] = None,
```

Comment:
This change, and all similar ones, will break python `< 3.9`. I think that `ruff` uses by default whatever version of python is available in the docker image (`3.10`, if I remember correctly), but there is a configuration option to prevent it from doing that. However, I don't remember which of the rules is enforcing this change (`pyupgrade` maybe?)  to check the actual name of this option.

Reply:
We should discuss deprecating 3.8, only has ~6 months of support left.

Reply:
I checked this. The target version is inferred from pyproject.toml.

`ruff --show-settings | grep  target_version`

gives

```target_version: Py38,
target_version: Py38,```